File: avro/avro-annotation-processor/src/main/java/io/koraframework/avro/annotation/processor/AvroReaderGenerator.java
package io.koraframework.avro.annotation.processor;

import com.palantir.javapoet.*;
import io.koraframework.annotation.processor.common.CommonClassNames;
import io.koraframework.annotation.processor.common.CommonUtils;
import org.jspecify.annotations.Nullable;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.element.TypeParameterElement;
import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;

public class AvroReaderGenerator {
    private final ProcessingEnvironment env;

    public AvroReaderGenerator(ProcessingEnvironment processingEnvironment) {
        this.env = processingEnvironment;
    }

    public void generate(TypeElement element) {
        var typeName = TypeName.get(element.asType());
        var typeBuilder = TypeSpec.classBuilder(AvroUtils.readerName(element))
            .addAnnotation(AnnotationSpec.builder(CommonClassNames.koraGenerated)
                .addMember("value", CodeBlock.of("$S", AvroReaderGenerator.class.getCanonicalName()))
                .build())
            .addAnnotation(AvroTypes.avro)
            .addSuperinterface(ParameterizedTypeName.get(AvroTypes.reader, typeName))
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addOriginatingElement(element);

        for (TypeParameterElement typeParameter : element.getTypeParameters()) {
            typeBuilder.addTypeVariable(TypeVariableName.get(typeParameter));
        }

        typeBuilder.addField(FieldSpec.builder(AvroTypes.schema, "SCHEMA")
            .addModifiers(Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
            .initializer("$T.getClassSchema()", typeName)
            .build());
        typeBuilder.addField(FieldSpec.builder(AvroTypes.specificData, "SPECIFIC_DATA")
            .addModifiers(Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
            .initializer("new $T().getSpecificData()", typeName)
            .build());
        typeBuilder.addField(FieldSpec.builder(ParameterizedTypeName.get(AvroTypes.datumReader, typeName), "READER")
            .addModifiers(Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
            .initializer("new $T<>(SCHEMA, SCHEMA, SPECIFIC_DATA)", AvroTypes.datumReader)
            .build());
        typeBuilder.addField(FieldSpec.builder(
                ParameterizedTypeName.get(AvroTypes.concurrentMap, AvroTypes.schema, ParameterizedTypeName.get(AvroTypes.datumReader, typeName)), "READERS_BY_WRITER_SCHEMA")
            .addModifiers(Modifier.PRIVATE, Modifier.FINAL)
            .initializer("new $T<>()", AvroTypes.concurrentHashMap)
            .build());

        typeBuilder.addMethod(MethodSpec.methodBuilder("getSchema")
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addAnnotation(Override.class)
            .returns(AvroTypes.schema)
            .addStatement("return SCHEMA")
            .build());

        var method = MethodSpec.methodBuilder("read")
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addAnnotation(Nullable.class)
            .addAnnotation(Override.class)
            .addParameter(TypeName.get(InputStream.class), "value")
            .returns(typeName);
        method.beginControlFlow("try");
        method.beginControlFlow("if (value == null || value.available() == 0)");
        method.addStatement("return null");
        method.endControlFlow();
        method.addStatement("var decoder = $T.get().directBinaryDecoder(value, null)", AvroTypes.decoderFactory);
        method.addStatement("return READER.read(new $T(), decoder)", typeName);
        method.nextControlFlow("catch ($T e)", IOException.class);
        method.addStatement("throw new $T(e)", UncheckedIOException.class);
        method.endControlFlow();

        typeBuilder.addMethod(method.build());

        var resolvingMethod = MethodSpec.methodBuilder("read")
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addAnnotation(Nullable.class)
            .addAnnotation(Override.class)
            .addParameter(ParameterSpec.builder(AvroTypes.schema, "writerSchema").addAnnotation(Nullable.class).build())
            .addParameter(TypeName.get(InputStream.class), "value")
            .returns(typeName);
        resolvingMethod.beginControlFlow("try");
        resolvingMethod.beginControlFlow("if (value == null || value.available() == 0)");
        resolvingMethod.addStatement("return null");
        resolvingMethod.endControlFlow();
        resolvingMethod.beginControlFlow("if (writerSchema == null || SCHEMA.equals(writerSchema))");
        resolvingMethod.addStatement("return read(value)");
        resolvingMethod.endControlFlow();
        resolvingMethod.addStatement("var reader = READERS_BY_WRITER_SCHEMA.computeIfAbsent(writerSchema, ws -> new $T<>(ws, SCHEMA, SPECIFIC_DATA))", AvroTypes.datumReader);
        resolvingMethod.addStatement("var decoder = $T.get().directBinaryDecoder(value, null)", AvroTypes.decoderFactory);
        resolvingMethod.addStatement("return reader.read(new $T(), decoder)", typeName);
        resolvingMethod.nextControlFlow("catch ($T e)", IOException.class);
        resolvingMethod.addStatement("throw new $T(e)", UncheckedIOException.class);
        resolvingMethod.endControlFlow();

        typeBuilder.addMethod(resolvingMethod.build());
        var javaFile = JavaFile.builder(AvroUtils.classPackage(this.env.getElementUtils(), element), typeBuilder.build()).build();
        CommonUtils.safeWriteTo(this.env, javaFile);
    }
}

File: avro/avro-symbol-processor/src/main/kotlin/io/koraframework/avro/symbol/processor/AvroReaderGenerator.kt
package io.koraframework.avro.symbol.processor

import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.addOriginatingKSFile
import com.squareup.kotlinpoet.ksp.toTypeParameterResolver
import com.squareup.kotlinpoet.ksp.toTypeVariableName
import com.squareup.kotlinpoet.ksp.writeTo
import io.koraframework.ksp.common.KspCommonUtils.generated
import io.koraframework.ksp.common.KspCommonUtils.toTypeName
import java.io.IOException
import java.io.InputStream
import java.io.UncheckedIOException

class AvroReaderGenerator(val resolver: Resolver, private val codeGenerator: CodeGenerator) {
    fun generate(declaration: KSClassDeclaration) {
        val typeName = declaration.toTypeName()
        val typeParameterResolver = declaration.typeParameters.toTypeParameterResolver()
        val readerInterface = AvroTypes.reader.parameterizedBy(typeName)
        val typeBuilder = TypeSpec.classBuilder(declaration.readerName())
            .generated(AvroReaderGenerator::class)
            .addAnnotation(AvroTypes.avro)
        declaration.containingFile?.let { typeBuilder.addOriginatingKSFile(it) }

        typeBuilder.addSuperinterface(readerInterface)
        declaration.typeParameters.forEach {
            typeBuilder.addTypeVariable(it.toTypeVariableName(typeParameterResolver))
        }
        typeBuilder.addProperty(PropertySpec.builder("SCHEMA", AvroTypes.schema).addModifiers(KModifier.PRIVATE, KModifier.FINAL).initializer("%T.getClassSchema()", typeName).build())
        typeBuilder.addProperty(PropertySpec.builder("SPECIFIC_DATA", AvroTypes.specificData).addModifiers(KModifier.PRIVATE, KModifier.FINAL).initializer("%T().getSpecificData()", typeName).build())
        typeBuilder.addProperty(PropertySpec.builder("READER", AvroTypes.datumReader.parameterizedBy(typeName)).addModifiers(KModifier.PRIVATE, KModifier.FINAL).initializer("%T(SCHEMA, SCHEMA, SPECIFIC_DATA)", AvroTypes.datumReader).build())
        typeBuilder.addProperty(
            PropertySpec.builder("READERS_BY_WRITER_SCHEMA", AvroTypes.concurrentMap.parameterizedBy(AvroTypes.schema, AvroTypes.datumReader.parameterizedBy(typeName)))
                .addModifiers(KModifier.PRIVATE, KModifier.FINAL)
                .initializer("%T()", AvroTypes.concurrentHashMap)
                .build()
        )

        typeBuilder.addFunction(
            FunSpec.builder("getSchema")
                .addModifiers(KModifier.PUBLIC, KModifier.FINAL, KModifier.OVERRIDE)
                .returns(AvroTypes.schema)
                .addStatement("return SCHEMA")
                .build()
        )

        val method = FunSpec.builder("read")
            .addModifiers(KModifier.PUBLIC, KModifier.FINAL, KModifier.OVERRIDE)
            .addParameter("value", InputStream::class.asTypeName().copy(true))
            .returns(typeName.copy(true))
        method.beginControlFlow("try")
        method.beginControlFlow("if (value == null || value.available() == 0)")
        method.addStatement("return null")
        method.endControlFlow()
        method.addStatement("val decoder = %T.get().directBinaryDecoder(value, null)", AvroTypes.decoderFactory)
        method.addStatement("return READER.read(%T(), decoder)", typeName)
        method.nextControlFlow("catch (e: %T)", IOException::class)
        method.addStatement("throw %T(e)", UncheckedIOException::class)
        method.endControlFlow()
        typeBuilder.addFunction(method.build())

        val resolvingMethod = FunSpec.builder("read")
            .addModifiers(KModifier.PUBLIC, KModifier.FINAL, KModifier.OVERRIDE)
            .addParameter("writerSchema", AvroTypes.schema.copy(true))
            .addParameter("value", InputStream::class.asTypeName().copy(true))
            .returns(typeName.copy(true))
        resolvingMethod.beginControlFlow("try")
        resolvingMethod.beginControlFlow("if (value == null || value.available() == 0)")
        resolvingMethod.addStatement("return null")
        resolvingMethod.endControlFlow()
        resolvingMethod.beginControlFlow("if (writerSchema == null || SCHEMA == writerSchema)")
        resolvingMethod.addStatement("return read(value)")
        resolvingMethod.endControlFlow()
        resolvingMethod.addStatement("val reader = READERS_BY_WRITER_SCHEMA.computeIfAbsent(writerSchema) { ws -> %T(ws, SCHEMA, SPECIFIC_DATA) }", AvroTypes.datumReader)
        resolvingMethod.addStatement("val decoder = %T.get().directBinaryDecoder(value, null)", AvroTypes.decoderFactory)
        resolvingMethod.addStatement("return reader.read(%T(), decoder)", typeName)
        resolvingMethod.nextControlFlow("catch (e: %T)", IOException::class)
        resolvingMethod.addStatement("throw %T(e)", UncheckedIOException::class)
        resolvingMethod.endControlFlow()
        typeBuilder.addFunction(resolvingMethod.build())

        val spec = typeBuilder.build()
        FileSpec.builder(classPackage(declaration), spec.name!!).addType(spec).build().writeTo(codeGenerator = codeGenerator, aggregating = false)
    }
}
